File: 234_PalindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# 55%
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        first, curr, pre, count = head, head, None, 0
        while curr:
            count += 1
            curr = curr.next
        count -= (count / 2)
        
        #Separate two parts
        curr = head
        while count:
            pre, curr, count = curr, curr.next, count - 1
        if pre: pre.next = None
        
        #Reverse second half part
        pre = None
        while curr:
            pre, curr.next, curr = curr, pre, curr.next
        
        #Compare two parts
        while first and pre:
            if first.val != pre.val:
                return False
            first, pre = first.next, pre.next
        return True
		
 # Better solution than mine      
 def isPalindrome(self, head):
    rev = None
    slow = fast = head
    while fast and fast.next:
        fast = fast.next.next
        rev, rev.next, slow = slow, rev, slow.next
    if fast:
        slow = slow.next
    while rev and rev.val == slow.val:
        slow = slow.next
        rev = rev.next
    return not rev      

# Same principle and keep input list as original
def isPalindrome(self, head):
    rev = None
    fast = head
    while fast and fast.next:
        fast = fast.next.next
        rev, rev.next, head = head, rev, head.next
    tail = head.next if fast else head
    isPali = True
    while rev:
        isPali = isPali and rev.val == tail.val
        head, head.next, rev = rev, head, rev.next
        tail = tail.next
    return isPali	

File: 234_PalindromeLinkedList.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
//72.22%
bool isPalindrome(struct ListNode* head) {
    struct ListNode *curr = head, *first = head;
	int count = 0;
	while(curr){
		count++;
		curr = curr -> next;
	}
	curr = head;
	count -= (count >> 1);
	struct ListNode *pre = NULL;
	
	//Split the list to two parts
	while(count--){
		pre = curr;
		curr = curr -> next;
	}
	if(pre)
		pre -> next = NULL;
	
	//Inverse second half list
	pre = NULL;
	while(curr){
		struct ListNode *next = curr -> next;
		curr -> next = pre;
		pre = curr;
		curr = next;
	}
	
	//Compare these two half parts started with 'pre' and 'first' separately
	while(pre && first){
		if(pre -> val != first -> val)
			return false;
		pre = pre -> next;
		first = first -> next;
	}
	return true;
}